Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -84,28 +78,25 @@ public function findByIdentifier($identifier)
     }
 
     /**
-     * @param  AggregateRootInterface $aggregate
+     * @param AggregateRootInterface $aggregate
+     * @param int $expectedVersion
      * @return void
      */
-    public function save(AggregateRootInterface $aggregate)
+    final public function save(AggregateRootInterface $aggregate, int $expectedVersion = null)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
-        try {
-            $stream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            $stream = new EventStream();
+        if (!$aggregate instanceof EventSourcedAggregateRootInterface) {
+            throw new \InvalidArgumentException(sprintf('This repository can only save instances implementing EventSourcedAggregateRootInterface, given: "%s"', get_class($aggregate)), 1477226552);
         }
-
+        $streamName = $this->streamNameResolver->getStreamNameForAggregate($aggregate);
         $uncommittedEvents = $aggregate->pullUncommittedEvents();
-        $stream->addEvents(...$uncommittedEvents);
+        $version = $aggregate->getVersion() + 1;
+        $uncommittedEventsWithMetadata = array_map(function(EventInterface $event) use(&$version) {
+            return new EventWithMetadata($event, new EventMetadata([Metadata::VERSION => $version ++]));
+        }, $uncommittedEvents);
 
-        $this->eventStore->commit($streamName, $stream, function ($version) use ($uncommittedEvents) {
-            /** @var EventTransport $eventTransport */
-            foreach ($uncommittedEvents as $eventTransport) {
-                // @todo metadata enrichment must be done in external service, with some middleware support
-                $versionedMetadata = $eventTransport->getMetadata()->withProperty(Metadata::VERSION, $version);
-                $this->eventBus->handle($eventTransport->withMetadata($versionedMetadata));
-            }
-        });
+        if ($expectedVersion === null) {
+            $expectedVersion = $aggregate->getVersion();
```

Comment:
Hm, didn't we decide to do optimistic concurrency check in the `findBy* `/loading of the aggregate? The check if the aggregate got changed during the runtime of this command can be done transparently (see above).

I think there's a difference between "User worked with an outdated aggregate" and "aggregate became outdated within this request":
- the first can not be safely resolved automatically in general, because the user decision was made on an already outdated view
- in the latter case, it should be safe to simply reattempt the command for this request a second time. The user decision was made on an up-to-date view at that point in time and things most likely didn't change dramatically - if they did in a way that conflicts, the second attempt will show that by the aggregate throwing an exception